smac_planner/src/a_star.cpp

```diff
@@ -220,62 +225,69 @@ bool AStarAlgorithm::createPath(IndexPath & path, int & iterations, const float
   return false;
 }
 
-bool AStarAlgorithm::isGoal(NodePtr & node)
+template<typename NodeT>
+bool AStarAlgorithm<NodeT>::isGoal(NodePtr & node)
 {
   return node == getGoal();
 }
 
-bool AStarAlgorithm::backtracePath(NodePtr & node, IndexPath & path)
+template<typename NodeT>
+bool AStarAlgorithm<NodeT>::backtracePath(NodePtr & node, IndexPath & path)
 {
-  if (!node->last_node) {
+  if (!node->parent) {
     return false;
   }
 
   NodePtr current_node = node;
 
-  while (current_node->last_node) {
+  while (current_node->parent) {
     path.push_back(current_node->getIndex());
-    current_node = current_node->last_node;
+    current_node = current_node->parent;
   }
 
   return path.size() > 1;
-
 }
 
-NodePtr & AStarAlgorithm::getStart()
+template<typename NodeT>
+typename AStarAlgorithm<NodeT>::NodePtr & AStarAlgorithm<NodeT>::getStart()
 {
   return _start;
 }
 
-NodePtr & AStarAlgorithm::getGoal()
+template<typename NodeT>
+typename AStarAlgorithm<NodeT>::NodePtr & AStarAlgorithm<NodeT>::getGoal()
 {
   return _goal;
 }
 
-NodePtr AStarAlgorithm::getNode()
+template<typename NodeT>
+typename AStarAlgorithm<NodeT>::NodePtr AStarAlgorithm<NodeT>::getNode()
```

Comment:
Shouldn't `NodePtr` work here? I don't think you need all of the other stuff since you define `NodePtr` in the header file

Reply:
Sadly it doesn't work just using `NodePtr` there, it gives a compilation error ` error: ‘NodePtr’ does not name a type;`. At that point the compiler doesn't know that `NodePtr` is a type, thus I have to use the fully qualified name. You can use `NodePtr` just fine within the methods, because at that point the compiler understands it is a `AStartAlgorithm<NodeT>`, but it doesn't work the same for return types (I didn't know this before, maybe I'm missing something.

There's a second option which is to re-define `NodePtr` at the beginning of the file with:
```
template<typename NodeT>
using NodePtr = NodeT *
```
And then we do
```
template<typename NodeT>
NodePtr<NodeT> AStartAlgorithm<NodeT>::getNode()
```
But I found this a bit less clean since it declares the same type twice.

Reply:
Ah, I see why, its because we have the NodePtr definition moved into the A* class. I didn't click that that move would have this impact, that works. I wonder if you can do a type def `using AStarAlgorithm<Node>::NodePtr Node2DPtr` or something to shorten this up.

I'm also fine leaving as is.